src/preact.d.ts

```diff
@@ -1,349 +1,414 @@
+		componentDidMount?(): void;
+		componentWillUnmount?(): void;
+		componentDidUnmount?(): void;
+		componentWillReceiveProps?(nextProps: Props, nextContext: any): void;
+		shouldComponentUpdate?(nextProps: Props, nextState: State, nextContext: any): boolean;
+		componentWillUpdate?(nextProps: Props, nextState: State, nextContext: any): void;
+		componentDidUpdate?(previousProps: Props, previousState: State, previousContext: any): void;
 	}
 
-	interface FunctionalComponent<PropsType> {
-		(props?: PropsType & ComponentProps<this>, context?: any): JSX.Element
+	/**
+	 *	@deprecated
+	 *	use StatelessComponent or SFC instead
+	 */
+	type FunctionalComponent<Props> = StatelessComponent<Props>;
+	type SFC<Props> = StatelessComponent<Props>;
+	interface StatelessComponent<Props> {
+		(props: Readonly<Props> & Readonly<ComponentProps>, context?: any): VNode<any>;
+		defaultProps?: Partial<Props>;
+		displayName?: string;
 	}
 
-	interface ComponentConstructor<PropsType, StateType> {
-		new (props?:PropsType, context?: any):Component<PropsType, StateType>;
+	/**
+	 *	@deprecated
+	 *	use ComponentClass instead
+	 */
+	type ComponentConstructor<Props, State> = ComponentClass<Props, State>;
+
+	interface ComponentClass<Props, State> {
+		new (props?: Props, context?: any): Component<Props, State>;
+		defaultProps?: Partial<Props>;
+		displayName?: string;
 	}
 
-	abstract class Component<PropsType, StateType> implements ComponentLifecycle<PropsType, StateType> {
-		constructor(props?:PropsType, context?:any);
+	// Type alias for a component considered generally, whether stateless or stateful.
+	type AnyComponent<Props, State> = SFC<Props> | ComponentClass<Props, State>;
 
-		static displayName?:string;
-		static defaultProps?:any;
+	abstract class Component<Props, State> implements ComponentLifecycle<Props, State> {
+		static displayName?: string;
+		// class property members cannot be typed via Generics
+		// -> so defaultProps wont match Props by inference, instead do:
+		// -- --> static defaultProps: Props = { myProp: 'hello' }
+		static defaultProps?: {};
 
-		state:StateType;
-		props:PropsType & ComponentProps<this>;
+		props: Readonly<ComponentProps> & Readonly<Props>;
+		state: Readonly<State>;
 		context: any;
-		base:HTMLElement;
-
-		linkState:(name:string) => (event: Event) => void;
+		base: HTMLElement;
 
-		setState<K extends keyof StateType>(state:Pick<StateType, K>, callback?:() => void):void;
-		setState<K extends keyof StateType>(fn:(prevState:StateType, props:PropsType) => Pick<StateType, K>, callback?:() => void):void;
+		// @TODO - preact-compat addition
+		// this is needed when used with react libs via preact-compat ( because React has refs on Component instance )
+		// refs?: any;
 
-		forceUpdate(): void;
+		constructor(props?: Props, context?: any);
 
-		abstract render(props?:PropsType & ComponentProps<this>, state?:StateType, context?: any):JSX.Element;
-	}
+		linkState: (name: string) => (event: Event) => void;
```

Comment:
Might want to mark this as deprecated, since (in Preact 8+) `linkState` is now in a separate module and not part of `Component` itself.